Add specialized addInput function to Accumulator - continuation

maciejgrzybek opened the pull request:
Supersedes https://github.com/prestodb/presto/pull/6793

Previously mentioned benchmarks reran with similar results.
In addition to them, I also benchmarked a few corner case use cases (relevant to simplifications used in code):
```sql
select count(x) from (select map_agg(partkey, tax) over(partition by orderkey) x from sf100.lineitem) v;
```
is ~8x faster
```sql
select count(x) from (select count(orderkey) over() x from sf100.lineitem) v;
```
is ~70x faster

They are special cases because in both cases blocks are copied using `Block.getSingleValueBlock()`.

The way I use `getSingleValueBlock()` could be illustrated with below pseudo-code snippet:
```java
state = ...
block = ...

for (int i = startPosition; i < endPosition; ++i) {
function.input(state, block.getSingleValueBlock(i), 0);
}
```
which I claim is correctness-wise equal to (which is current implementation from `master`):

```java
state = ...
block = ...

for (int i = startPosition; i < endPosition; ++i) {
function.input(state, block, i);
}
```

---

This addresses https://github.com/prestodb/presto/pull/6793#issuecomment-268896263.

There's obviously a room for further improvement (to not copy blocks using `Block.getSingleValueBlock()` per position for `BLOCK_CHANNEL_INPUT`) but this patch improves a lot already, therefore I think it's worth to merge this patch now and leave further improvements (which may not be straight forward to implement) for later or when it's needed.

maciejgrzybek commented:
@electrum: have you had a chance to glance over it?

maciejgrzybek commented:
Rebased on master.

maciejgrzybek commented:
Rebased on master.

maciejgrzybek commented:
Rebased on master with bunch of fixups:
1. included initialization of `isNull` variable into `loopBody` (https://github.com/prestodb/presto/pull/7190/commits/6aa846a74efb921b9dc2778a91fbf2948efb0a39#diff-63141fd2addf84640b3cd1cb0e3b9af6R274)
2. Made `isNull` `boolean` instead of `int` (https://github.com/prestodb/presto/pull/7190/commits/ddc7604334657364c5219029201c6876d287e056#diff-63141fd2addf84640b3cd1cb0e3b9af6R260)
3. Refactored `generateReadFromIndex` to reuse defined `Variables` instead of retrieving them from `scope` (https://github.com/prestodb/presto/pull/7190/commits/2bd9e15e39601f194fcd499f425e4e905d717457#diff-63141fd2addf84640b3cd1cb0e3b9af6R440).

---

@electrum / @haozhun Will you have a chance to review it any time soon-ish?

maciejgrzybek commented:
@electrum: We've had this PR open since December last year. I understand it's not easy to properly review this PR given it's code gen. Is there anything we can do to help you with this review? Maybe throw someone from us to review it to build up your confidence in this PR?
We can also talk about it on Hangout or Slack. I can walk you through the code, if you wish.

maciejgrzybek commented:
After 3 weeks without response after last question...
Can we help somehow to make this PR easier to review for you, guys?
CC: @martint

dain commented:
I started looking at this and a few others last night.

maciejgrzybek commented:
Thanks! Looking forward to hear any feedback!

dain commented:
Review status:
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/158bb58b9f80668c7d3d0bc6973bfdf8eb41ca9b Enable Accumulator to retrieve a block of single value from the index
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/ddc7604334657364c5219029201c6876d287e056 Add Accumulator.addInput(WindowIndex)
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/8c99e531bafee745e4ac38c290d0031dd8a85695 Add Accumulator.addInput for window functions without parameters
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/83da46e87623296d0cd1eb61e2087a5ed22d4621 Enable PagesWindowIndex retrieving Object stored as Block
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/d577b066e2ab3ab4e917574d25964edb89305a57 Add test for map_agg in window context
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/6aa846a74efb921b9dc2778a91fbf2948efb0a39 Accept range of positions in addInput(WindowIndex)
- [ ]  https://github.com/prestodb/presto/pull/7190/commits/2bd9e15e39601f194fcd499f425e4e905d717457 Accept list of channels in Accumulator

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/function/WindowIndex.java:
```diff
@@ -72,6 +73,22 @@
 
     /**
+     * Gets a value stored as a {@link Block}.
+     * @param channel argument number
```
Newline between description and first `@param`. Fill in `@return` description or remove it.

dain commented on presto-spi/src/main/java/com/facebook/presto/spi/function/WindowIndex.java:
```diff
@@ -81,6 +81,14 @@
+     * @param position row within the partition, starting at zero
+     * @return
+     */
+    Block getObject(int channel, int position);
```
This should return type `Object` to match the `Pages.getObject`.  In the long run the plan is to support any object type from these method.

maciejgrzybek commented:
@dain: Thanks for extensive description. I like your proposed simplification!
I was not aware of `BytecodeExpressions`.
I was simply trying to mimic existing (in the rest of `AccumulatorCompiler`) `BytecodeNode`-style code.

I squashed the code as you requested and rewritten as you suggested (with a few minor changes).
Please have a look at https://github.com/prestodb/presto/pull/7775.
